source/common/shared_pool/shared_pool.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <thread>
+#include <type_traits>
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/singleton/instance.h"
+
+#include "common/common/assert.h"
+#include "common/common/non_copyable.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace SharedPool {
+
+/**
+ * Used to share objects that have the same content.
+ * control the life cycle of shared objects by reference counting
+ * Note: this class has only deleteObject for thread safety, and the other member methods are
+ * non-thread safe
```

Comment:
Do you mind adding a bit more comments here in the ownership model? And possibly more comments below? This is neat and makes sense to me but it took a few read throughs to understand which thread is doing what, the use of weak/strong pointers, etc.

Reply:
I add some comments, but I don't know if it's enough for someone else to read.